File: app/src/main/java/com/savvasdalkitsis/gameframe/gameframe/usecase/GameFrameUseCase.java
package com.savvasdalkitsis.gameframe.gameframe.usecase;

import android.net.wifi.WifiManager;
import android.support.annotation.NonNull;
import android.text.format.Formatter;
import android.util.Log;

import com.savvasdalkitsis.gameframe.gameframe.api.CommandResponse;
import com.savvasdalkitsis.gameframe.gameframe.api.GameFrameApi;
import com.savvasdalkitsis.gameframe.ip.model.IpAddress;
import com.savvasdalkitsis.gameframe.ip.model.IpNotFoundException;
import com.savvasdalkitsis.gameframe.ip.usecase.IpDiscoveryUseCase;
import com.savvasdalkitsis.gameframe.model.Brightness;
import com.savvasdalkitsis.gameframe.model.ClockFace;
import com.savvasdalkitsis.gameframe.model.CycleInterval;
import com.savvasdalkitsis.gameframe.model.DisplayMode;
import com.savvasdalkitsis.gameframe.model.PlaybackMode;

import java.io.File;
import java.io.IOException;
import java.util.Map;

import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import rx.Observable;
import rx.functions.Func1;

import static com.savvasdalkitsis.gameframe.ip.model.IpAddress.Builder.ipAddress;
import static java.util.Collections.singletonMap;

public class GameFrameUseCase {

    private final OkHttpClient okHttpClient;
    private final WifiManager wifiManager;
    private final GameFrameApi gameFrameApi;
    private final IpDiscoveryUseCase ipDiscoveryUseCase;

    public GameFrameUseCase(OkHttpClient okHttpClient, WifiManager wifiManager, GameFrameApi gameFrameApi, IpDiscoveryUseCase ipDiscoveryUseCase) {
        this.okHttpClient = okHttpClient;
        this.wifiManager = wifiManager;
        this.gameFrameApi = gameFrameApi;
        this.ipDiscoveryUseCase = ipDiscoveryUseCase;
    }

    public Observable<Void> togglePower() {
        return gameFrameApi.command(param("power")).compose(mapResponse());
    }

    public Observable<Void> menu() {
        return gameFrameApi.command(param("menu")).compose(mapResponse());
    }

    public Observable<Void> next() {
        return gameFrameApi.command(param("next")).compose(mapResponse());
    }

    public Observable<Void> setBrightness(Brightness brightness) {
        return gameFrameApi.set(param(brightness.getQueryParamName()));
    }

    public Observable<Void> setPlaybackMode(PlaybackMode playbackMode) {
        return gameFrameApi.set(param(playbackMode.getQueryParamName()));
    }

    public Observable<Void> setCycleInterval(CycleInterval cycleInterval) {
        return gameFrameApi.set(param(cycleInterval.getQueryParamName()));
    }

    public Observable<Void> setDisplayMode(DisplayMode displayMode) {
        return gameFrameApi.set(param(displayMode.getQueryParamName()));
    }

    public Observable<Void> setClockFace(ClockFace clockFace) {
        return gameFrameApi.set(param(clockFace.getQueryParamName()));
    }

    public Observable<CreateFolderResponse> createFolder(String name) {
        return gameFrameApi.command(singletonMap("mkdir", name))
                .flatMap(response -> {
                    if (isSuccess(response)) {
                        return Observable.just(CreateFolderResponse.SUCCESS);
                    } else if (response.getMessage().contains("exists")) {
                        return Observable.just(CreateFolderResponse.ALREADY_EXISTS);
                    } else {
                        return Observable.error(wrap(response));
                    }
                });
    }

    public Observable<Void> removeFolder(String name) {
        return gameFrameApi.command(singletonMap("rmdir", name)).compose(mapResponse());
    }

    public Observable<Void> uploadFile(File file) {
        RequestBody requestFile = RequestBody.create(MediaType.parse("image/bmp"), file);
        MultipartBody.Part filePart = MultipartBody.Part.createFormData("my_file[]", "0.bmp", requestFile);
        return gameFrameApi.upload(filePart).compose(mapResponse());
    }

    public Observable<IpAddress> discoverGameFrameIp() {
        return Observable.defer(this::getDeviceIp)
                .flatMap(wholePart4Subrange())
                .concatMap(ip -> {
                    ipDiscoveryUseCase.emitMonitoredAddress(ip);
                    try {
                        if (isFromGameFrame(ping(ip))) {
                            return Observable.just(ip);
                        }
                    } catch (IOException e) {
                        Log.w(GameFrameUseCase.class.getName(), "Error trying to call " + ip, e);
                    }
                    return Observable.empty();
                })
                .first()
                .onErrorResumeNext(e -> Observable.error(new IpNotFoundException("Game Frame IP not found", e)));
    }

    private Func1<IpAddress, Observable<? extends IpAddress>> wholePart4Subrange() {
        return ip -> Observable.range(0, 256)
                .map(String::valueOf)
                .map(part -> ipAddress(ip).part4(part).build());
    }

    @SuppressWarnings("deprecation")
    private Observable<IpAddress> getDeviceIp() {
        String ip = Formatter.formatIpAddress(wifiManager.getConnectionInfo().getIpAddress());
        return Observable.just(IpAddress.parse(ip));
    }

    private boolean isFromGameFrame(Response response) {
        return response.isSuccessful()
                && response.headers().get("Server").startsWith("Webduino/");
    }

    private Response ping(IpAddress ip) throws IOException {
        return okHttpClient.newCall(new Request.Builder()
                .url("http://" + ip.toString() + "/command")
                .post(RequestBody.create(MediaType.parse("application/x-www-form-urlencoded"), "next="))
                .build()).execute();
    }

    @NonNull
    private Map<String, String> param(String name) {
        return singletonMap(name, "");
    }

    private Observable.Transformer<CommandResponse, Void> mapResponse() {
        return o -> o
                .flatMap(response -> {
                    if (isSuccess(response)) {
                        return Observable.just(null);
                    } else {
                        return Observable.error(wrap(response));
                    }
                });
    }

    private GameFrameCommandError wrap(CommandResponse response) {
        return new GameFrameCommandError("Command was not successful. response: " + response);
    }

    private boolean isSuccess(CommandResponse response) {
        return "status".equals(response.getStatus());
    }
}

File: app/src/main/java/com/savvasdalkitsis/gameframe/gameframe/usecase/GameFrameUseCase.kt
package com.savvasdalkitsis.gameframe.gameframe.usecase

import android.net.wifi.WifiManager
import android.text.format.Formatter
import android.util.Log
import com.savvasdalkitsis.gameframe.control.model.*
import com.savvasdalkitsis.gameframe.gameframe.api.CommandResponse
import com.savvasdalkitsis.gameframe.gameframe.api.GameFrameApi
import com.savvasdalkitsis.gameframe.gameframe.model.AlreadyExistsOnGameFrameException
import com.savvasdalkitsis.gameframe.ip.model.IpAddress
import com.savvasdalkitsis.gameframe.ip.model.IpNotFoundException
import com.savvasdalkitsis.gameframe.ip.usecase.IpDiscoveryUseCase
import io.reactivex.Completable
import io.reactivex.Flowable
import io.reactivex.Single
import io.reactivex.functions.Function
import okhttp3.*
import java.io.File
import java.io.IOException
import java.util.Collections.singletonMap

class GameFrameUseCase(private val okHttpClient: OkHttpClient,
                       private val wifiManager: WifiManager,
                       private val gameFrameApi: GameFrameApi,
                       private val ipDiscoveryUseCase: IpDiscoveryUseCase) {

    fun togglePower(): Completable =
            gameFrameApi.command(param("power")).to(mapResponse())

    fun menu(): Completable =
            gameFrameApi.command(param("menu")).to(mapResponse())

    fun next(): Completable  =
            gameFrameApi.command(param("next")).to(mapResponse())

    fun setBrightness(brightness: Brightness) =
            gameFrameApi.set(param(brightness.queryParamName))

    fun setPlaybackMode(playbackMode: PlaybackMode) =
            gameFrameApi.set(param(playbackMode.queryParamName))

    fun setCycleInterval(cycleInterval: CycleInterval) =
            gameFrameApi.set(param(cycleInterval.queryParamName))

    fun setDisplayMode(displayMode: DisplayMode) =
            gameFrameApi.set(param(displayMode.queryParamName))

    fun setClockFace(clockFace: ClockFace) =
            gameFrameApi.set(param(clockFace.queryParamName))

    fun createFolder(name: String): Completable  =
            gameFrameApi.command(singletonMap("mkdir", name))
                    .flatMapCompletable { response -> when {
                        isSuccess(response) ->
                            Completable.complete()
                        response.message?.contains("exists") ?: false ->
                            Completable.error(AlreadyExistsOnGameFrameException("Could not create folder on game frame with name '$name' as it already exists"))
                        else ->
                            Completable.error(wrap(response))
                    } }

    fun removeFolder(name: String): Completable =
            gameFrameApi.command(singletonMap("rmdir", name)).to(mapResponse())

    fun uploadFile(file: File): Completable {
        val requestFile = RequestBody.create(MediaType.parse("image/bmp"), file)
        val filePart = MultipartBody.Part.createFormData("my_file[]", "0.bmp", requestFile)
        return gameFrameApi.upload(filePart).to(mapResponse())
    }

    fun play(name: String): Completable =
            gameFrameApi.command(singletonMap("play", name)).to(mapResponse())

    fun discoverGameFrameIp(): Single<IpAddress> {
        return Single.defer(this::deviceIp)
                .flattenAsFlowable(wholePart4Subrange())
                .concatMap { ip ->
                    ipDiscoveryUseCase.emitMonitoredAddress(ip)
                    var result: Flowable<IpAddress> = Flowable.empty<IpAddress>()
                    try {
                        if (isFromGameFrame(ping(ip))) {
                            result = Flowable.just(ip)
                        }
                    } catch (e: IOException) {
                        Log.w(GameFrameUseCase::class.java.name, "Error trying to call " + ip, e)
                    }
                    result
                }
                .firstOrError()
                .onErrorResumeNext { e -> Single.error<IpAddress>(IpNotFoundException("Game Frame IP not found", e)) }
    }

    private fun wholePart4Subrange() = { ip: IpAddress ->
        (0..256).map { it.toString() }.map { part -> ip.copy(part4 = part) }
    }

    private val deviceIp: Single<IpAddress>
        get() {
            @Suppress("DEPRECATION")
            val ip = Formatter.formatIpAddress(wifiManager.connectionInfo.ipAddress)
            return Single.defer { Single.just(IpAddress.parse(ip)) }
        }

    private fun isFromGameFrame(response: Response) =
            response.isSuccessful && response.headers().get("Server")?.startsWith("Webduino/") ?: false

    @Throws(IOException::class)
    private fun ping(ip: IpAddress) = okHttpClient.newCall(Request.Builder()
            .url("http://" + ip.toString() + "/command")
            .post(RequestBody.create(MediaType.parse("application/x-www-form-urlencoded"), "next="))
            .build()).execute()

    private fun param(name: String) = singletonMap(name, "")

    private fun mapResponse(): Function<Single<CommandResponse>, Completable> = Function { s ->
        s.flatMapCompletable {
            when {
                isSuccess(it) -> Completable.complete()
                else -> Completable.error(wrap(it))
            }
        }
    }

    private fun wrap(response: CommandResponse?) =
            GameFrameCommandError("Command was not successful. response: " + response!!)

    private fun isSuccess(response: CommandResponse?) = "success" == response?.status
}
